pom.xml

```diff
@@ -130,6 +130,12 @@
       <version>1.1.1</version>
       <scope>test</scope>
     </dependency>
+    <!-- https://mvnrepository.com/artifact/org.jenkins-ci.plugins/trilead-api -->
+    <dependency>
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>trilead-api</artifactId>
+      <version>1.0.1</version>
```

Comment:
I think this cannot work so long as your baseline is 2.138.x. This version of `trilead-api` requires 2.150.x.

Reply:
Maybe you meant to use 1.0.0? But I am not even sure if things will work naturally when you are running on any line prior to 2.190.x, as Jenkins core will be bundling Trilead so this dependency would at best be ignored (at worst might mix some different versions of classes in the same package).

Reply:
Maybe I'm misreading the pom for trilead-api, but I think it [declares jenkins.version as 2.89.4](https://github.com/jenkinsci/trilead-api-plugin/blob/trilead-api-1.0.1/pom.xml#L50).

I'll configure a quick test with Jenkins 2.138 LTS to see if I can confirm that it passes some interactive smoke tests.

Reply:
Hmm, https://github.com/jenkinsci/trilead-api-plugin/commit/a71565d9ec349ae65426eb700c0500728019eb6d and https://github.com/jenkinsci/trilead-api-plugin/commit/0fa9517d2795dc4e1f98fe8c980fa1383f7de112 do not agree. There was a bump to 2.150.1 which occurred prior to one. Did @kuisathaverat botch some releases? Anyway I checked in http://repo.jenkins-ci.org/public/org/jenkins-ci/plugins/trilead-api/1.0.1/ and indeed it seems to be on 2.89.x. Still,

> Jenkins core will be bundling Trilead so this dependency would at best be ignored (at worst might mix some different versions of classes in the same package).

Reply:
@jglick a prototype of a merge of this change to the master branch has been running various authentication tests successfully on multiple agents connected to a Jenkins 2.138.4 master with latest plugin versions supported on that release.

That may only indicate that the tested configurations are just lucky or that I didn't test enough to detect the root problem.  Any other recommendations of things to check?

Unfortunately, if this technique is not workable, I'm not sure what path to take to allow plugin compatibility tester to run with git client plugin on Jenkins 2.190.1 while still supporting older versions.  Any recommendations?

Reply:
Since my test installation of 2.138.4 already had trilead-api 1.0.1 installed (I assume in order to satisfy some dependency in the ssh-slaves plugin), it seems to me like a reasonably safe assumption that trilead-api 1.0.1 is available and will be used in the 2.138.4 environment.

Reply:
I suspect it was not being used by `ssh-slaves`, but fine.

Reply:
trilead-api 1.0.2 was screw up so I've had to release 1.0.3 minutes after 1.0.2

Reply:
The trilead-api plugin 1.0.0 is used by [ssh-slaves 1.29.1](https://github.com/jenkinsci/ssh-slaves-plugin/blob/ssh-slaves-1.29.1/pom.xml#L101).  It was released Nov 2018. That is the version of ssh-slaves in my test installation that is running Jenkins 2.138.4.

Version 1.0.1 of the trilead-api plugin is used by [ssh-slaves 1.29.0](https://github.com/jenkinsci/ssh-slaves-plugin/blob/ssh-slaves-1.29.0/pom.xml#L101), also released in Nov 2018.

If someone is using an ssh-slaves plugin from a year ago or newer, doesn't that assure us that the trilead-api plugin is already installed in their environment?

If this is not the preferred compromise to allow plugin compatibility tester to run on Jenkins 2.190.1 with the git client plugin, what alternatives should I be investigating?

Reply:
ssh-slaves 1.29.1 was another epic fail three days later I've had to release [ssh-slaves-1.29.2](https://github.com/jenkinsci/ssh-slaves-plugin/blob/ssh-slaves-1.29.2/pom.xml) there the trilead-api is used only on test, it is still used only on test, so ssh-slaves relay on the trilead-ssh2 lib provided by the core right now. When 2.190 would release I could release an ssh-slaves plugin that depends on trilead-api 1.0.5

Reply:
Thanks @kuisathaverat for the insights on specific versions of ssh-slaves.  I've been the very happy recipient of your improvements to the ssh-slaves plugin.  They have made my life easier, especially since I can now manage my Windows agents using ssh.

Your comment seems to support the concern from @jglick that depending on trilead-api 1.0.1 or trilead-api 1.0.0 may not be the best solution to allow us to run a single version of git client plugin in plugin compatibility tester for Jenkins 2.176.4 and for Jenkins 2.190.1.

Any suggestions of alternate approaches?  

We could release a git client plugin with minimum Jenkins version 2.190.1, but I think that may surprise or confuse users of earlier Jenkins versions.  

We could list the failures as an exception in the Jenkins 2.190.1 plugin compatibility tests, but that seems like it hides the issue rather than resolving it.

Reply:
I have the same issue with ssh-slaves, I think the only solution is to bump the minimum core to 1.190.1, I do not want to mess with the Jenkins class loader, I never have luck with it.

Reply:
> We could list the failures as an exception in the Jenkins 2.190.1 plugin compatibility tests

That is what I may end up doing in https://github.com/jenkinsci/bom/pull/110, unfortunately. Still need to work out how exactly to move that forward.

> it hides the issue rather than resolving it

Well…yes, though the “issue” may merely be that PCT is complaining about nothing. The whole system of `JenkinsRule` + PCT has fundamental limitations. See [JENKINS-41827](https://issues.jenkins-ci.org/browse/JENKINS-41827) for discussion. PCT does attempt to honor plugin splits and perhaps I just need to fix the BOM build.

Personally my inclination is generally to just keep it simple and bump the Jenkins baseline. Critical bug fixes are [easy enough](https://gist.github.com/jglick/86a30894446ed38f918050c1180483e2) to backport to an older line. If some user of an older line complains about being unable to get the latest minor fixes or features in some plugin, you can point out that they are running a version of Jenkins with acknowledged security vulnerabilities.